compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIESplitLogicalObjectFifos.cpp

```diff
@@ -0,0 +1,214 @@
+    return WalkResult::advance();
+  });
+
+  if (l2ToL1DmaOps.size() == 0) return;
+
+  SmallVector<OpFoldResult> baseSourceOffsets =
+      l2ToL1DmaOps[0].getSourceMixedOffsets();
+  DenseSet<unsigned> splitDimensionsSet;
+  for (unsigned i = 1, n = l2ToL1DmaOps.size(); i < n; i++) {
+    SmallVector<OpFoldResult> sourceOffsets =
+        l2ToL1DmaOps[i].getSourceMixedOffsets();
+    for (unsigned j = 0, m = baseSourceOffsets.size(); j < m; j++) {
+      if (baseSourceOffsets[j] != sourceOffsets[j]) {
+        splitDimensionsSet.insert(j);
+      }
+    }
+  }
+
+  OpFoldResult zeroVal = getAsIndexOpFoldResult(context, 0);
+  OpFoldResult oneVal = getAsIndexOpFoldResult(context, 1);
+  DenseSet<Operation *> toBeErased;
+  for (AMDAIE::DmaCpyNdOp l2ToL1DmaOp : l2ToL1DmaOps) {
+    LogicalObjectFifoFromMemrefOp sourceObjectFifo =
+        l2ToL1DmaOp.getSourceObjectFifo();
+    auto sourceAllocOp =
+        sourceObjectFifo.getMemref().getDefiningOp<memref::AllocOp>();
+    uint64_t sourceMemrefSpace = sourceObjectFifo.getMemorySpaceAsUInt();
+    if (!sourceAllocOp || sourceMemrefSpace != 1) continue;
+    LogicalObjectFifoFromMemrefOp targetObjectFifo =
+        l2ToL1DmaOp.getTargetObjectFifo();
+    Value targetAllocOp = targetObjectFifo.getMemref();
+
+    // Fetch the L3 -> L2 Dma Op corresponding to the L2 buffer as target.
+    AMDAIE::DmaCpyNdOp l3ToL2DmaOp;
+    for (Operation *objFifoUserOp : sourceObjectFifo->getUsers()) {
+      if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(objFifoUserOp);
+          dmaOp.getTargetObjectFifo() == sourceObjectFifo) {
+        l3ToL2DmaOp = dmaOp;
+        toBeErased.insert(dmaOp);
+        break;
+      }
+    }
+    toBeErased.insert(sourceAllocOp);
+    toBeErased.insert(sourceObjectFifo);
+
+    SmallVector<OpFoldResult, 6> staticL2AsSourceOffsets =
+        l2ToL1DmaOp.getSourceMixedOffsets();
+    SmallVector<OpFoldResult, 6> staticL2AsSourceSizes =
+        l2ToL1DmaOp.getSourceMixedSizes();
+    SmallVector<OpFoldResult, 6> staticL2AsSourceStrides =
+        l2ToL1DmaOp.getSourceMixedStrides();
+    SmallVector<OpFoldResult, 4> staticL2AsTargetOffsets =
+        l3ToL2DmaOp.getTargetMixedOffsets();
+    SmallVector<OpFoldResult, 4> staticL2AsTargetSizes =
+        l3ToL2DmaOp.getTargetMixedSizes();
+    SmallVector<OpFoldResult, 4> staticL2AsTargetStrides =
+        l3ToL2DmaOp.getTargetMixedStrides();
+    SmallVector<int64_t, 4> l2ShapeAsTarget = llvm::to_vector(
+        cast<MemRefType>(
+            l3ToL2DmaOp.getTargetObjectFifo().getMemref().getType())
+            .getShape());
+    DenseMap<int64_t, int64_t> dimToOffsetMapForL3AsSource;
+    for (unsigned dim : splitDimensionsSet) {
+      std::optional<int64_t> constantOffset =
+          getConstantIntValue(staticL2AsSourceOffsets[dim]);
+      if (!constantOffset) {
+        l2ToL1DmaOp->emitOpError("found a non-constant value for offset");
+        return signalPassFailure();
```

Comment:
Should we error out? Can't we just return if the offsets are non-constant and maybe add a debug message? And this could also be checked earlier in this function.

Reply:
I've used `emitRemark` instead. Let me know if it looks okay to you now?

I've not preponed this check and chose to keep it here because we need to check the same for L3->L2 and anyway need to use those constants at this part of the function for computing the new offset.